### PR TITLE
Full switch to elm 0.19

### LINKED ISSUE #8: Support elm 0.19; "elm-make" is now "elm make"
`elm-make` has been absorbed in to the elm compiler, and is now executed as `elm make` (so `make` is the first argument for the `elm` command).

This means I now have to create an alias `elm-make` that translates back to `elm make`. Can the project please be updated to use `elm` as the command and `'make'` as the first argument to `self.subprocess()`?

E.g.

```python
    write_args = [elm, 'make', source, '--output', tmp.name]
```

Note that the `--yes` command-line switch is now gone.

It would also be nice if we could switch on `--optimize` or `--debug`, either as explicit options or via a general `ELM_EXTRA_ARGUMENTS` config option.

<sub>*Edited to cover all 0.19 changes*</sub>

### PR DESCRIPTION
**Pull request by mjpieters**
This is perhaps a big change, but adding support for elm 0.19 and keeping backward compatibility with `elm-make` (elm <= 0.18) is not worth the effort in my opinion.

Instead, those that still need to work with the older elm releases could just pin `webassets-elm` to a previous release.

This would fully resolve #8

### CODE REVIEW

File: .travis.yml
Comment (cuducos) [inline .travis.yml]: Python 3.4 has also reached its [end of life in March](https://www.python.org/downloads/release/python-3410/), let's drop support for them?

File: CHANGELOG
Comment (cuducos) [inline CHANGELOG]: I would rephrase that to reinforce that we added support for something and dropped support for others:


>  * Added support for Elm 0.19 (including `--optimize` and `--debug` options)
>  * Drop support for Elm 0.18 and older Elm versions
>  * Drop support for Python 3.3 and 3.4

What do you think?

File: README.rst
Comment (cuducos) [inline README.rst]:
```suggestion
* `ELM_OPTIMIZE`: enable the Elm compiler optimization option. Recommended for production output.
```

File: README.rst
Comment (cuducos) [inline README.rst]:
```suggestion
* `ELM_DEBUG`: enable the Elm compiler debug option.
```

File: setup.py
Comment (cuducos) [inline setup.py]: Just a reminder that if we drop support for 3.4 (as suggested earlier), we need to remove this line too ; )

File: setup.py
Comment (cuducos) [inline setup.py]: We need to have a Python 3.7 line here, don't we?

File: tox.ini
Comment (cuducos) [inline tox.ini]: Just a reminder that if we drop support for 3.4 (as suggested earlier), we need to remove it here too ; )

File: webassets_elm/__init__.py
Comment (cuducos) [inline webassets_elm/__init__.py]: I think the `_BIN` suffix already tells the user it's the binary, and the parenthesis as used before are mostly a type hint, so what about:

```suggestion
    ELM_BIN (string, path to the Elm binary)
```

File: webassets_elm/__init__.py
Comment (cuducos) [inline webassets_elm/__init__.py]:
```suggestion
        If set to True, turns on the Elm compiler optimizations.
```

File: webassets_elm/__init__.py
Comment (cuducos) [inline webassets_elm/__init__.py]:
```suggestion
        If set to True, turns on the Elm compiler time traveling debugger option.
```

File: webassets_elm/__init__.py
Comment (cuducos) [inline webassets_elm/__init__.py]:
```suggestion
        (https://github.com/elm-lang/elm-make/issues/177), so we need to
```

File: webassets_elm/__init__.py
Comment (cuducos) [inline webassets_elm/__init__.py]: I was wondering… is there a situation in which `or False` would evaluate to `True`? I mean, this could be simplified to:

```suggestion
        if self.optimize:
```

File: webassets_elm/__init__.py
Comment (cuducos) [inline webassets_elm/__init__.py]: Surely I was also wondering… is there a situation in which `or False` would evaluate to `True` here too? I mean, this could be simplified to:

```suggestion
        if self.debug:
```

File: webassets_elm/__init__.py
Comment (cuducos) [inline webassets_elm/__init__.py]: Can we have more readable variable/object names? `tempd` could be `directory`, for example.

File: webassets_elm/__init__.py
Comment (cuducos) [inline webassets_elm/__init__.py]: Can we have more readable variable/object names? `outf` could be `output`, `js`, `output_file`, `compiled_path` for example.

File: webassets_elm/__init__.py
Comment (cuducos) [inline webassets_elm/__init__.py]: As you have used the `list` methods earlier (e.g. `args.append('--debug')` instead of `args + ['--debug']`) what about keeping this pattern here?

```suggestion
            args .extend(['--output', outf])
```

File: webassets_elm/__init__.py
Comment (cuducos) [inline webassets_elm/__init__.py]: What does `inf` means here? Can we have a more meaningful variable name?

File: webassets_elm/__init__.py
Comment (cuducos) [inline webassets_elm/__init__.py:2]: As we're only using `copyfileobj`, maybe we could consider importing `from shutil import copyfileobj`…

File: webassets_elm/tests.py
Comment (cuducos) [inline webassets_elm/tests.py:25]: This is a JSON content and I think there are advantages in keeping it as a separate file (for example: linters and visual hints in the code editor when reviewing or maintaining it).

Can we step back and use an external file as it was before **or** (dirty workaround disclaimer) keep it here as a Python dictionary and use `json.dumps` when we need the content in JSON as string?

File: webassets_elm/tests.py
Comment (cuducos) [inline webassets_elm/tests.py]: Would it make sense to assert we haven't used `--optimize` and `--debug` here?

**Review by cuducos — CHANGES_REQUESTED** [review]
Many thanks for this PR, @mjpieters! I do think it's a great contribution and I agree that keeping backward compatibility is too complex: Elm 0.18 users (or older) can pin the version.

Before merging and publishing  I would like to consider a set of really small changes, and discuss them with you. Besides what I added as inline comments, what do you think about:

* Adding in the `README.rst` a notice telling about Elm 0.19 (and suggesting the last version offering support to older versions)?
* Maybe capturing the error if one tries to run your PR with Elm 0.18 and set an error message helping the user to understand that the current versions requires Elm 0.19 and if the user needs to work with older Elm versions, suggest the proper version to pin?
* Clarifying the rational to drop `ELM_MAKE_CHANGE_DIRECTORY`? I got the drop of the custom context manager, but not the drop of the user option.
* Would you be considering using/adding [Black](https://github.com/python/black) as a default in this package? It could be just a check in the CI by now.

Let me know if you're willing to work on these issues I raised, otherwise feel free to just share your thoughts and I can take from here.

File: .travis.yml
Comment (mjpieters) [reply .travis.yml]: This is your project, but yes, I'd drop support for 3.4. :-)

File: webassets_elm/__init__.py
Comment (mjpieters) [reply webassets_elm/__init__.py]: This was patterned on the existing webassets documentation, see the [`less` filter](https://webassets.readthedocs.io/en/latest/builtin_filters.html#webassets.filter.less.Less), and [`less_ruby`](https://webassets.readthedocs.io/en/latest/builtin_filters.html#less-ruby). Seeing this again now (both are basically the same plugin), I agree with you more than that precedent.

I'll just remove that postfix and leave it to the next line to explain the setting better, like it's done for the [`compass` plugin](https://webassets.readthedocs.io/en/latest/builtin_filters.html#webassets.filter.compass.Compass).

File: webassets_elm/__init__.py
Comment (mjpieters) [reply webassets_elm/__init__.py]: Nope, that's a bit silly that one. I was tired when I wrote those.

File: webassets_elm/__init__.py
Comment (mjpieters) [reply webassets_elm/__init__.py]: `compilation_result` and further down, `compilation_result_file`.

File: webassets_elm/__init__.py
Comment (mjpieters) [reply webassets_elm/__init__.py]: I prefer `+= [...]`, but you are right, this doesn't need to be a new variable.

File: webassets_elm/__init__.py
Comment (mjpieters) [reply webassets_elm/__init__.py:2]: Personally, I prefer to import the module, it documents the context better when you use `shutil.copyfileobj()`.

File: webassets_elm/tests.py
Comment (mjpieters) [reply webassets_elm/tests.py:25]: This data is only needed by the Elm compiler *during testing*, as is the input file. There is no reason to have this kick around as a separate file in the project, and actually serves to confuse if it does live on the filesystem (*there's an `elm.json` file, but no Elm source files?*). So I strongly prefer to keep this as close to the tests as possible.

In the same vein, we don't need any access to the contents from Python here, this is an opaque blob only needed to make the integration with the compiler work. And since the `TempEnvironmentHelper` base class expects strings in the `default_files` mapping, if you do store this as a Python dictionary you'd have to serialise to JSON first.

So I don't see any advantages to changing this back to a file on disk, nor to storing it as a Python dict.

File: webassets_elm/tests.py
Comment (mjpieters) [reply webassets_elm/tests.py]: Sure, we can add a test for the warning Elm puts in there. I've added the same to the other 2 tests, so we now test for the presence or absence of the two markers, `Compiled in DEV mode`, and `Compiled in DEBUG mode`.

**Comment by mjpieters** [comment]
I've addressed the inline comments. I've still to address the README, to which I agree. Sorry, out of time for now to do those, hopefully tomorrow.

**Comment by mjpieters** [comment]
Addressing your other comments:

> * Maybe capturing the error if one tries to run your PR with Elm 0.18 and set an error message helping the user to understand that the current versions requires Elm 0.19 and if the user needs to work with older Elm versions, suggest the proper version to pin?

I am not certain how to do that in a clean way and should perhaps be a separate issue altogether. Is there a `elm` command when using 0.18? Would detecting 0.18 involve having to run `elm --version` each time? Is that added overhead worth it?

> * Clarifying the rational to drop `ELM_MAKE_CHANGE_DIRECTORY`? I got the drop of the custom context manager, but not the drop of the user option.

Because it doesn't break existing project layouts, and makes more work when enabled. The `elm make` command will find a `elm.json` file in parent directories, but not in child directories. Moving the current working dir to that of the `.elm` source file should always result in finding the parent `elm.json` file, but not doing this will break for projects that put the whole Elm source tree in a subdir. I do this for the current project, for example.

Fewer options makes a plugin easier to use.

> * Would you be considering using/adding Black as a default in this package? It could be just a check in the CI by now.

I would very much love to use Black. I recommend using a pre-commit hook, I use https://pre-commit.com/ which [Black supports directly](https://github.com/python/black#version-control-integration). I've applied Black to this pull request, I'll leave adding the hook to you.

**Comment by mjpieters** [comment]
I've addressed all that I can address, I think. Anything else that needs changing?

**Comment by cuducos** [comment]
Many many thanks for all the effort, @mjpieters : ) That is a great moment for this tiny package ❤️ 

I'm gonna merge this to a separate branch because I cannot agree with two points you raised:

* the removal of `ELM_MAKE_CHANGE_DIRECTORY` (I do maintain some projects that have Elm in sub-directories, as the repo is organized around the backend, so we have a a specific sub-directory for the front-end)
* keeping JSON contents as Python strings — that is an anti-pattern in my opinion, something that just make maintenance more difficult (I'm gonna move the Elm code to a separate file as well)

Once I align these issues and review the docs, I'll release a new version to PyPI (hopefully today).

**Comment by mjpieters** [comment]
Not a problem.

One note on

> * the removal of `ELM_MAKE_CHANGE_DIRECTORY` (I do maintain some projects that have Elm in sub-directories, as the repo is organized around the backend, so we have a a specific sub-directory for the front-end)

That's exactly what I have currently; `assets/elm` contains the Elm front-end code. The current-working-dir is changed to that of the source file being compiled (in my case in `assets/elm/src`) and everything works great.

Note that I replaced the old default of **not** changing directories (`ELM_MAKE_CHANGE_DIRECTORY=False`), to *always* changing directories (so `ELM_MAKE_CHANGE_DIRECTORY=True`), perhaps that was missed?

**Comment by cuducos** [comment]
Oops! My bad. Forget what I've said about `ELM_MAKE_CHANGE_DIRECTORY`! Now I fully understand what changed in Elm 0.19 compiler ; )

**Pull request merged**
